Azure storage filesystem support

spacedragon opened the pull request:
```
export AZURE_STORAGE_KEY=<storage key>
tritonserver --model-repository=as://<storage account>.blob.core.windows.net/<container>/<path>
```

Commit:
azure storage filesystem

omnipo commented on src/core/filesystem.cc:
```diff
@@ -681,6 +690,351 @@ GCSFileSystem::WriteTextFile(
+  {
+    return Status(Status::Code::INTERNAL, "blob client initialize failed.");
+  }
+  as::blob_client_wrapper bc(client_);
```
is this not used?

spacedragon replied:
I will remove it.

omnipo commented on src/core/filesystem.cc:
```diff
@@ -681,6 +690,351 @@ GCSFileSystem::WriteTextFile(
+
+    account = std::make_shared<as::storage_account>(
+        account_name, cred, /* use_https */ true);
+    client_ = std::make_shared<as::blob_client>(account, 16);
```
what is this '16'?

spacedragon replied:
max_concurrency

adding comments.

omnipo commented on src/core/filesystem.cc:
```diff
@@ -681,6 +690,351 @@ GCSFileSystem::WriteTextFile(
+      account_name = host_name.substr(0, pos);
+    } else {
+      account_name = host_name;
+    }
```
Does this convention have a doc?

spacedragon replied:
No. `{account}.blob.core.windows.net` is the default assigned DNS name. But user can assign any domain to its storage address.

Adding a new env `AZURE_STORAGE_ACCOUNT` to cover this situation.

omnipo commented on src/core/filesystem.cc:
```diff
@@ -681,6 +690,351 @@ GCSFileSystem::WriteTextFile(
+  std::string host_name, container, blob_path, query;
+  if (RE2::FullMatch(
+          path, as_regex_, &host_name, &container, &blob_path, &query)) {
+    size_t pos = host_name.rfind(".blob.core.windows.net");
```
how is rfind better than find?

omnipo commented on src/core/filesystem.cc:
```diff
@@ -681,6 +690,351 @@ GCSFileSystem::WriteTextFile(
+  std::ostringstream out_stream;
+  bc.download_blob_to_stream(container, object_path, 0, 0, out_stream);
+  if (errno!=0)
+  {
```
let's follow the coding style here and move this brace to the line above

omnipo commented on src/core/filesystem.cc:
```diff
@@ -681,6 +690,351 @@ GCSFileSystem::WriteTextFile(
+    return Status(Status::Code::INTERNAL, "Failed check file: " + path);
+  }
+  if (blobs.blobs.size() > 0) {
+        *exists = true;
```
nit: indent

omnipo commented on src/core/filesystem.cc:
```diff
@@ -1163,6 +1517,21 @@ GetFileSystem(const std::string& path, FileSystem** file_system)
 
+  // Check if this is an Azure Storage path 
+  if (!path.empty() && !path.rfind("as://", 0)) {
+#ifndef TRITON_ENABLE_S3
```
Should be ENABLE_AZURE?

omnipo commented on src/core/filesystem.cc:
```diff
@@ -681,6 +690,351 @@ GCSFileSystem::WriteTextFile(
+      time_t last_modified;
+      bc.download_blob_to_file(container, blob_path, local_path, last_modified);
+      if (errno != 0) 
+      {
```
brace :)

Commit:
Include python3 and python_backend dependencies in server image

Commit:
Update documentation to be github resident (#2102)

Commit:
Documentation is centralized in docs/ directory, so remove library README

Commit:
L0_docs is removed so don't attempt to get models from there

aleksa2808 committed:
Fix typo in client_libraries.md (#2106)

rotorliu committed:
Fix opencv4 error (#2071)

Co-authored-by: NAVINFO\liubao <liubao@navinfo.com>

mvpel committed:
Add multi-output ability, model name argument (#2093)

Allow ensemble_image_client.py to request, receive, and display multiple outputs from a model, and add the -m / --model-name option to specify a model to use for the inference.

Tabrizian committed:
Fix non-graceful test (#2108)

Commit:
Add --github-organization flag to build.py

Commit:
Remove --backend-config flag documentation to tensorflow_backend repo

Commit:
qa test for azure storage

Commit:
Compile core to export server and backend API symbols (#2114)

CoderHam committed:
Add L0_infer_tf2 directory (#2113)

* Add L0_infer_tf2 directory

* Add check to L0_infer_tf2 for unset TEST_VALGRIND

Commit:
fix redfine maco compile error

Commit:
fixing sas

Commit:
remove logs

Commit:
change 'as' to 'azure_storage'

Commit:
Merge branch 'yulong/azure_storage' of https://github.com/spacedragon/triton-inference-server into yulong/azure_storage

Commit:
Merge branch 'master' of https://github.com/spacedragon/triton-inference-server into yulong/azure_storage

Commit:
remove sas support
